### PR TITLE
relates to #947: added validations for NamespacesResource

### PR DESCRIPTION
**Pull request by ivansenic**
Same as before, this time for the `NamespacesResource`:

* validate body, removed unused and unnecessary code
* added DTO for the creation with validation and extracted getting the `Replication`
* moved all tests from the `BaseDocumentApiV2Test` to a new class, with well defined structure, added more tests for validation

I know this heavily intersects with the `v2/schemas/keyspaces`, but the ticket was related to the docs api resource validation, we can merge as part of #101.

### CODE REVIEW

**Comment by ivansenic** [comment]
I am getting a strange query execution error on C* 4 and DSE, with the query `CREATE KEYSPACE IF NOT EXISTS "YUz4wAanzFziOk3z" WITH replication = { 'class' : 'NetworkTopologyStrategy', 'dc2': 3, 'dc1': 1 }`. Works on C* 3.11. Can anybody advise?

```
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,066 ProcessRunner.java:78 - stargate17-0> INFO  [MigrationStage:1] 2021-05-24 16:09:42,065 NetworkTopologyStrategy.java:84 - Configured datacenter replicas are dc2:rf(3), dc1:rf(1)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> ERROR [dw-206 - POST /v2/schemas/namespaces] 2021-05-24 16:09:42,069 RequestHandler.java:93 - Error when executing request
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> java.util.concurrent.CompletionException: org.apache.cassandra.stargate.exceptions.ConfigurationException: Unrecognized strategy option {dc2} passed to NetworkTopologyStrategy for keyspace SbtehbrQNMj1ZsnX
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:661)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:646)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at io.stargate.db.cassandra.impl.CassandraPersistence.lambda$runOnExecutor$1(CassandraPersistence.java:266)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:162)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$LocalSessionFutureTask.run(AbstractLocalAwareExecutorService.java:134)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:119)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at java.lang.Thread.run(Thread.java:748)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> Caused by: org.apache.cassandra.stargate.exceptions.ConfigurationException: Unrecognized strategy option {dc2} passed to NetworkTopologyStrategy for keyspace SbtehbrQNMj1ZsnX
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at io.stargate.db.cassandra.impl.Conversion.toExternal(Conversion.java:323)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at io.stargate.db.cassandra.impl.Conversion.convertInternalException(Conversion.java:451)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at io.stargate.db.cassandra.impl.CassandraPersistence$CassandraConnection.lambda$executeRequestOnExecutor$0(CassandraPersistence.java:450)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	at io.stargate.db.cassandra.impl.CassandraPersistence.lambda$runOnExecutor$1(CassandraPersistence.java:257)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	... 6 common frames omitted
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 	Suppressed: org.apache.cassandra.exceptions.ConfigurationException: Unrecognized strategy option {dc2} passed to NetworkTopologyStrategy for keyspace SbtehbrQNMj1ZsnX
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.locator.AbstractReplicationStrategy.validateExpectedOptions(AbstractReplicationStrategy.java:455)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.locator.NetworkTopologyStrategy.validateExpectedOptions(NetworkTopologyStrategy.java:291)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.locator.AbstractReplicationStrategy.validateReplicationStrategy(AbstractReplicationStrategy.java:401)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.schema.ReplicationParams.validate(ReplicationParams.java:78)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.schema.KeyspaceParams.validate(KeyspaceParams.java:94)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.cql3.statements.schema.CreateKeyspaceStatement.apply(CreateKeyspaceStatement.java:77)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,070 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.schema.Schema.transform(Schema.java:587)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,071 ProcessRunner.java:78 - stargate17-0> 		at org.apache.cassandra.schema.MigrationManager.lambda$announce$2(MigrationManager.java:221)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,071 ProcessRunner.java:78 - stargate17-0> 		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,071 ProcessRunner.java:78 - stargate17-0> 		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,071 ProcessRunner.java:78 - stargate17-0> 		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,071 ProcessRunner.java:78 - stargate17-0> 		... 2 common frames omitted
Step #4 - "cassandra-4.0": INFO  [Exec Stream Pumper] 2021-05-24 16:09:42,072 ProcessRunner.java:78 - stargate17-0> INFO  [dw-206] 2021-05-24 16:09:42,072 Slf4jRequestLogWriter.java:62 - 127.0.0.1 - - [24/May/2021:16:09:42 +0000] "POST /v2/schemas/namespaces HTTP/1.1" 500 206 "-" "okhttp/4.9.0" 55
Step #4 - "cassandra-4.0": ERROR [main] 2021-05-24 16:09:42,076 RestUtils.java:279 - {"description":"Server error: org.apache.cassandra.stargate.exceptions.ConfigurationException: Unrecognized strategy option {dc2} passed to NetworkTopologyStrategy for keyspace SbtehbrQNMj1ZsnX","code":500}
Step #4 - "cassandra-4.0": WARN  [main] 2021-05-24 16:09:42,076 ExternalStorage.java:140 - Error in [engine:junit-jupiter]/[class:io.stargate.it.http.docsapi.NamespaceResourceIntTest]/[nested-class:CreateNamespace]/[method:networkTopologyStrategy()]
Step #4 - "cassandra-4.0": org.opentest4j.AssertionFailedError: 
Step #4 - "cassandra-4.0": Expecting:
Step #4 - "cassandra-4.0":  <500>
Step #4 - "cassandra-4.0": to be equal to:
Step #4 - "cassandra-4.0":  <201>
Step #4 - "cassandra-4.0": but was not.
Step #4 - "cassandra-4.0": 	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
Step #4 - "cassandra-4.0": 	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
Step #4 - "cassandra-4.0": 	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
Step #4 - "cassandra-4.0": 	at io.stargate.it.http.RestUtils.assertStatusCode(RestUtils.java:276)
Step #4 - "cassandra-4.0": 	at io.stargate.it.http.RestUtils.postWithHeader(RestUtils.java:88)
Step #4 - "cassandra-4.0": 	at io.stargate.it.http.RestUtils.post(RestUtils.java:99)
Step #4 - "cassandra-4.0": 	at io.stargate.it.http.docsapi.NamespaceResourceIntTest$CreateNamespace.networkTopologyStrategy(NamespaceResourceIntTest.java:208)
```

File: restapi/src/main/java/io/stargate/web/docsapi/models/dto/CreateNamespace.java
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/docsapi/models/dto/CreateNamespace.java:63]: Not a big deal, but could `replicas` and `datacenter` be optional types instead of converting them?

https://github.com/FasterXML/jackson-modules-java8

File: restapi/src/main/java/io/stargate/web/docsapi/models/dto/UpgradeCollection.java
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/docsapi/models/dto/UpgradeCollection.java:46]: Nice catches with the naming.

**Review by mpenick — APPROVED** [review]
+1

File: restapi/src/main/java/io/stargate/web/docsapi/models/dto/CreateNamespace.java
Comment (ivansenic) [reply restapi/src/main/java/io/stargate/web/docsapi/models/dto/CreateNamespace.java:63]: Idea is good, I tried, but seems that validations are not working anymore for anything wrapped in the optional..

File: testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java
Comment (dimas-b) [inline testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java:2002]: Could we delay this change until #937 is merged, please? :)

**Review by dimas-b — CHANGES_REQUESTED** [review]

**Review by dimas-b — COMMENTED** [review]
LGTM :+1: pending rebasing test changes on top of #937

File: testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java
Comment (ivansenic) [reply testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java:2002]: Sure, waiting :clock1:

Comment (ivansenic) [reply testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java:1988]: @dimas-b need your approval to merge :pray:

Comment (dimas-b) [reply testing/src/main/java/io/stargate/it/http/docsapi/BaseDocumentApiV2Test.java:1988]: sorry, forgot about this yesterday.... approving

**Review by dimas-b — APPROVED** [review]

**Pull request merged**
